### PR TITLE
HZB first mip (0) copy, resolution limit in 4096 pixels

### PR DESCRIPTION
**Pull request by BurgheRlyeh**

### CODE REVIEW

File: Saber/DeferredShadingComputeShader.hlsl [hunk 1]
```diff
@@ -67,35 +67,9 @@ float2 BestUVDerivative(
     return uvBest == float2(1.f, 1.f) ? float2(0.f, 0.f) : uvBest;
```
Comment (AMSpivak) [inline Saber/DeferredShadingComputeShader.hlsl:67]: Good idea to use it here

File: Saber/DepthBuffer.cpp [hunk 2]
```diff
@@ -63,35 +64,52 @@ void DepthBuffer::Resize(
...
 	);
-	ThrowIfFailed(pDevice->GetDeviceRemovedReason());
 
-	// Hierarchical Z-Buffer
+	ResizeHZB(pDevice, pAllocator, width, height);
+}
+
+void DepthBuffer::ResizeHZB(Microsoft::WRL::ComPtr<ID3D12Device2> pDevice, Microsoft::WRL::ComPtr<D3D12MA::Allocator> pAllocator, UINT64 width, UINT height) {
+	if (!m_pSinglePassDownsampler) {
+		return;
```
Comment (AMSpivak) [inline Saber/DepthBuffer.cpp]: if you use early return, maybe return code error or at least true  / false on function toleave possibility process success/ fail?

File: Saber/DepthBuffer.cpp [hunk 3]
```diff
@@ -117,6 +140,10 @@ D3D12_GPU_DESCRIPTOR_HANDLE DepthBuffer::GetSrvGpuDescHandle() const {
 	return m_pSrvsRange->GetGpuHandle();
 }
 
+D3D12_GPU_DESCRIPTOR_HANDLE DepthBuffer::GetSrvGpuDescHandleWithMips() const {
+	return m_pSrvsRange->GetGpuHandle(1);
```
Comment (AMSpivak) [inline Saber/DepthBuffer.cpp:181]: better constexpr variables with understandable names. or enum, but nod magic values

File: Saber/DepthBuffer.h [hunk 4]
```diff
@@ -43,15 +50,109 @@ class DepthBuffer {
...
+	);
+
+	void CreateHierarchicalDepthBuffer(
+		Microsoft::WRL::ComPtr<ID3D12GraphicsCommandList2> pCommandList,
+		Microsoft::WRL::ComPtr<ID3D12DescriptorHeap> pDescHeap
+	) {
+		if (!m_pSinglePassDownsampler) {
+			return;
+		}
+		//ResourceTransition(pCommandList, m_pDepthBuffer->GetResource(),
```
Comment (AMSpivak) [inline Saber/DepthBuffer.h]: why so many transitions?  do we need them?

File: Saber/DepthBuffer.h [hunk 5]
```diff
@@ -43,15 +50,109 @@ class DepthBuffer {
...
+		//	D3D12_RESOURCE_STATE_UNORDERED_ACCESS
+		//);
+		//pCommandList->CopyTextureRegion(
+		//	&CD3DX12_TEXTURE_COPY_LOCATION(m_pHZBuffer->GetResource().Get(), 0),
+		//	0, 0, 0,
+		//	&CD3DX12_TEXTURE_COPY_LOCATION(m_pDepthBuffer->GetResource().Get(), 0),
+		//	&CD3DX12_BOX(0, 0, 0, m_width, m_height, 1)
+		//);
+		
+		pCommandList->CopyTextureRegion(
```
Comment (AMSpivak) [inline Saber/DepthBuffer.h]: copy will be on compute queue?

File: Saber/Renderer.cpp [hunk 6]
```diff
@@ -364,8 +362,8 @@ void Renderer::PerformResize() {
         return;
 
     // Don't allow 0 size swap chain back buffers.
-    m_clientWidth = std::max(1u, width);
-    m_clientHeight = std::max(1u, height);
+    m_clientWidth  = std::min<uint32_t>(std::max(1u, width), 4096);
```
Comment (AMSpivak) [inline Saber/Renderer.cpp:366]: so - no 8k gaming& leave as is now, later add second spd pass for this case, maybe 16k gaming will usethis engine too

File: Saber/SinglePassDownsampler.h [hunk 7]
```diff
@@ -72,5 +72,76 @@ class SinglePassDownsampler : public ComputeObject {
 private:
     static Microsoft::WRL::ComPtr<ID3DBlob> CreateRootSignatureBlob(
         Microsoft::WRL::ComPtr<ID3D12Device2> pDevice
-    );
+    ) {
```
Comment (AMSpivak) [inline Saber/SinglePassDownsampler.h]: What is the reason to keep code in header instead of cpp?

File: Saber/SinglePassDownsamplerCS.hlsl [hunk 8]
```diff
@@ -87,6 +86,41 @@ void SpdStoreIntermediate(uint x, uint y, float4 value)
     spdIntermediate[x][y] = value;
 }
 
+// copy texture
```
Comment (AMSpivak) [inline Saber/SinglePassDownsamplerCS.hlsl]: are we doing this or not?

**Review by AMSpivak — CHANGES_REQUESTED** [review]

File: Saber/SinglePassDownsampler.h (hunk 7 above)
Comment (BurgheRlyeh) [reply Saber/SinglePassDownsampler.h]: no reason, moved

File: Saber/DepthBuffer.cpp (hunk 2 above)
Comment (BurgheRlyeh) [reply Saber/DepthBuffer.cpp]: fixed, now using bool return value, also added for >4096 resolution

File: Saber/DepthBuffer.cpp (hunk 3 above)
Comment (BurgheRlyeh) [reply Saber/DepthBuffer.cpp:181]: added indices for srvs

File: Saber/DepthBuffer.h (hunk 4 above)
Comment (BurgheRlyeh) [reply Saber/DepthBuffer.h]: Removed

File: Saber/DepthBuffer.h (hunk 5 above)
Comment (BurgheRlyeh) [reply Saber/DepthBuffer.h]: Everything on direct queue during rendering

File: Saber/Renderer.cpp (hunk 6 above)
Comment (BurgheRlyeh) [reply Saber/Renderer.cpp:366]: :( Also TODO in DepthBuffer::ResizeHZB

File: Saber/SinglePassDownsamplerCS.hlsl (hunk 8 above)
Comment (BurgheRlyeh) [reply Saber/SinglePassDownsamplerCS.hlsl]: just unsuccessful attempts, removed

**Review by AMSpivak — APPROVED** [review]

**Pull request merged**